stdlib/public/core/StringUTF8Validation.swift

```diff
@@ -0,0 +1,169 @@
+private func _isUTF8MultiByteLeading(_ x: UInt8) -> Bool {
+  return (0xC2...0xF4).contains(x)
+}
+
+private func _isNotOverlong_F0(_ x: UInt8) -> Bool {
+  return (0x90...0xBF).contains(x)
+}
+
+private func _isNotOverlong_F4(_ x: UInt8) -> Bool {
+  return _isContinuation(x) && x <= 0x8F
+}
+
+private func _isNotOverlong_E0(_ x: UInt8) -> Bool {
+  return (0xA0...0xBF).contains(x)
+}
+
+private func _isNotOverlong_ED(_ x: UInt8) -> Bool {
+  return _isContinuation(x) && x <= 0x9F
+}
+
+private func _isASCII_cmp(_ x: UInt8) -> Bool {
+  return x <= 0x7F
+}
+
+@usableFromInline
+struct UTF8ExtraInfo: Equatable {
+  public var isASCII: Bool
+}
+
+@_frozen
+@usableFromInline
+enum UTF8ValidationResult {
+  case success(UTF8ExtraInfo)
+  case error(toBeReplaced: Range<Int>)
+}
+
+extension UTF8ValidationResult: Equatable {}
+
+private struct UTF8ValidationError: Error {}
+
+@usableFromInline
+func validateUTF8(_ buf: UnsafeBufferPointer<UInt8>) -> UTF8ValidationResult {
+  var iter = buf.makeIterator()
+  var lastValidIndex: UInt = UInt(buf.startIndex) // UInt so we don't need to overflow check
+
+  @inline(__always) func guaranteeIn(_ f: (UInt8) -> Bool) throws {
+    guard let cu = iter.next() else { throw UTF8ValidationError() }
+    guard f(cu) else { throw UTF8ValidationError() }
+  }
+  @inline(__always) func guaranteeContinuation() throws {
+    try guaranteeIn(_isContinuation)
+  }
+
+  func findInvalidRange(_ buf: Slice<UnsafeBufferPointer<UInt8>>) -> Range<Int> {
+    var endIndex = buf.startIndex
+    var iter = buf.makeIterator()
+    _ = iter.next()
+    while let cu = iter.next(), !_isASCII(cu) && !_isUTF8MultiByteLeading(cu) {
+      endIndex += 1
+    }
+    let illegalRange = Range(buf.startIndex...endIndex)
+    _sanityCheck(illegalRange.clamped(to: (buf.startIndex..<buf.endIndex)) == illegalRange,
+                 "illegal range out of full range")
+    return illegalRange
+  }
+
+  do {
+    var isASCII = true
+    while let cu = iter.next() {
+      if _isASCII(cu) { lastValidIndex &+= 1; continue }
+      isASCII = false
+      if _slowPath(!_isUTF8MultiByteLeading(cu)) {
+        throw UTF8ValidationError()
+      }
+      // TODO: perhaps a custom branching pattern
+      switch cu {
+      case 0xC2...0xDF:
+        try guaranteeContinuation()
+        lastValidIndex &+= 2
```

Comment:
note for @milseman : this is new in this version: no longer checking overflow here, therefore made `lastValidIndex` a `UInt`, so overflow check moves down (in only one place now, where we make that an `Int`)

Reply:
Ok cool, just didn’t want to introduce extra restrictions but more than happy to make it `Int`. Will do tomorrow when I’m back at a computer